sympy/integrals/meijerint.py

```diff
@@ -1660,6 +1661,18 @@ def _meijerint_indefinite_1(f, x):
         fac_ = fac * C / (b*a**((1 + c)/b))
         rho = (c + 1)/b - 1
 
+        # handle zero denominators in the above substitution
+        special_results = []
+        for den in (a, b):
+            if not den.is_Number and not den.is_nonzero:
+                zero_sol = solve(den, dict=True)
```

Comment:
I'm not sure if we should worry about doing solve here. This seems like something that piecewise simplification should worry about. Do the other integration methods do this when they split into piecewise?

Reply:
I tried something like this instead of solve (diff against the first two commits):

```
diff --git a/sympy/integrals/meijerint.py b/sympy/integrals/meijerint.py
index 36609255ee..c72a78963f 100644
--- a/sympy/integrals/meijerint.py
+++ b/sympy/integrals/meijerint.py
@@ -1665,13 +1665,16 @@ def _meijerint_indefinite_1(f, x):
         special_results = []
         for den in (a, b):
             if not den.is_Number and den.is_zero is not False:
-                zero_sol = solve(den, dict=True)
-                for sol in zero_sol:
-                    zero_res = integrate(f.subs(sol), x)
-                    zero_cond = cond
-                    for vz, sz in sol.items():
-                        zero_cond = And(zero_cond, Eq(vz, sz))
-                    special_results.append(Piecewise((zero_res, zero_cond)))
+                _, den_mul_args = den.as_coeff_mul()
+                den = Mul(*den_mul_args)
+                f_replaced = f.xreplace({den: 0})
+                if f_replaced == f:
+                    # leave unevaluated
+                    zero_res = Integral(f_replaced, x)
+                else:
+                    zero_res = integrate(f_replaced, x)
+                zero_cond = And(cond, Eq(den, 0))
+                special_results.append(Piecewise((zero_res, zero_cond)))
 
         # we now use t**rho*G(params, t) = G(params + rho, t)
         # [L, page 150, equation (4)]

```

which works for this case:

```
>>> x, n, k = symbols("x, n, k") 
>>> integrate(sin(x**(n*k)), x)
Piecewise((x*sin(1), Eq(k*n, 0)), (x*x**(k*n)*gamma(1/2 + 1/(2*k*n))*hyper((1/2 + 1/(2*k*n),), (3/2, 3/2 + 1/(2*k*n)), -x**(2*k*n)/4)/(2*k*n*gamma(3/2 + 1/(2*k*n))), True))
```

but leaves the special case unevaluated with this example:

```
>>> integrate(sin(x**(n*k + 1)), x)
Piecewise((Integral(sin(x*x**(k*n)), x), Eq(2*k*n + 2, 0)), (x*x**(k*n)*x**(k*n/(k*n + 1))*x**(1/(k*n + 1))*gamma(1/2 + 1/(2*(k*n + 1)))*hyper((1/2 + 1/(2*(k*n + 1)),), (3/2, 3/2 + 1/(2*(k*n + 1))), -x**2*x**(2*k*n)/4)/(2*k*n*gamma(3/2 + 1/(2*(k*n + 1))) + 2*gamma(3/2 + 1/(2*(k*n + 1)))), True))
```

whereas with solve:

```
>>> integrate(sin(x**(n*k + 1)), x)
Piecewise((x*sin(1), Eq(k, -1/n)), (x*x**(k*n)*x**(k*n/(k*n + 1))*x**(1/(k*n + 1))*gamma(1/2 + 1/(2*(k*n + 1)))*hyper((1/2 + 1/(2*(k*n + 1)),), (3/2, 3/2 + 1/(2*(k*n + 1))), -x**2*x**(2*k*n)/4)/(2*k*n*gamma(3/2 + 1/(2*(k*n + 1))) + 2*gamma(3/2 + 1/(2*(k*n + 1)))), True))
```

I think heurisch is doing something similar with solve here:
https://github.com/sympy/sympy/blob/a44cc429c0c38e9abf6e60d421a94bb2f1de26c9/sympy/integrals/heurisch.py#L146

Reply:
I see. It needs to call solve so that it can get a chance to actually evaluate the integral.